tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift

```diff
@@ -0,0 +1,184 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LinuxSystemHeaders
+
+class ElfFile {
+  public enum Error: Swift.Error {
+    case FileOpenFailure(_ filePath: String)
+    case FileReadFailure(_ filePath: String, offset: UInt64, size: UInt64)
+    case FileNotElfFormat(_ filePath: String)
+    case MalformedElfFile(_ filePath: String, description: String = "")
+  }
+
+  public typealias SymbolMap = [String: (start: UInt64, end: UInt64)]
+
+  let filePath: String
+  let file: FileHandle
+  let ehdr: ElfEhdr
+  let isElf64: Bool
+
+  public init(filePath: String) throws {
+    self.filePath = filePath
+
+    guard let file = try? FileHandle(forReadingFrom: URL(fileURLWithPath: filePath)) else {
+      throw Error.FileOpenFailure(filePath)
+    }
```

Comment:
Does it make sense to use `Data(contentsOf:options:)` to use a mmap'ed view?

Reply:
I can if you want that, but I prefer explicit file i/o over `mmap` because it allows explicit error handling. They both share pages in the page cache so isn't going to perform significantly better than the other.